pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,851 @@
+        machine. We check if the user has passed user_data and the type of
+        image_id we are receiving, which can be snapshots ids or not.
+
+        Args:
+            vm_name: string, The name of the virtual machine.
+            image_id: string, The identifier of an image.
+            nic_id: string, The network interface id.
+            user_data: string, The user data to be passed to the
+                       virtual machine.
+
+        Returns:
+            A dict containing the parameters to provision a virtual machine.
+
+        """
+        vm_parameters = {
+            "location": self.location,
+            "hardware_profile": {
+                "vm_size": "Standard_DS1_v2"
+            },
+            "storage_profile": {
+                "image_reference": {}
+            },
+            "os_profile": {
+                "computer_name": vm_name,
+                "admin_username": self.username,
+                "linux_configuration": {
+                    "ssh": {
+                        "public_keys": [
+                            {
+                                "path": "/home/{}/.ssh/authorized_keys".format(
+                                    self.username),
+                                "key_data": self.key_pair.public_key_content
+                            }
+                        ]
+                    },
+                    "disable_password_authentication": True
+                }
+            },
+            "network_profile": {
+                "network_interfaces": [{
+                    "id": nic_id,
+                }]
+            },
+            "tags": {
+                "name": self.tag
+            }
+        }
+
+        if user_data:
+            # We need to encode the user_data into base64 before sending
+            # it to the virtual machine.
+            vm_parameters["os_profile"]["custom_data"] = base64.b64encode(
+                user_data.encode()).decode()
+
+        # We have two types of images, base images that are found on
+        # Azure marketplace and custom images based on other virtual
+        # machines. For marketplace images, we have an expected
+        # format: Canonical:UbuntuServer:19.10-DAILY. Here we have
+        # the published, offer and sku information. However, the
+        # snapshot images are just an id, similar to any other
+        # resource id that we find in Azure. Because of that
+        # difference, we need to treat those image types
+        # differently.
+        if image_id.startswith('Canonical'):
+            vm_parameters["storage_profile"]["image_reference"] = {
+                "publisher": 'Canonical',
+                "offer": util.get_offer_from_image_id(image_id),
+                "sku": util.get_sku_from_image_id(image_id),
+                "version": "latest"
+            }
+        else:
+            vm_parameters["storage_profile"]["image_reference"] = {
+                "id": image_id
+            }
```

Comment:
Maybe it's worth creating a private method for this here 
_get_image_reference_params(image_id)

This could return your dict
either 
{publisher:.... offer:... sku:... version:...} 
OR
{'id': image_id}